.github/workflows/run-test-cases.yml

```diff
@@ -88,211 +88,223 @@ jobs:
+            version: v1.18.9+k3s1
+            crictl: v1.17.0
+          - runtime: K3s-1.19
+            version: v1.19.4+k3s1
+            crictl: v1.17.0
+          - runtime: K3s-1.20
+            version: v1.20.0+k3s2
+            crictl: v1.17.0
+          - runtime: Kubernetes-1.16
+            version: 1.16.15-00
+            crictl: UNUSED
+          - runtime: Kubernetes-1.17
+            version: 1.17.14-00
+            crictl: UNUSED
+          - runtime: Kubernetes-1.18
+            version: 1.18.12-00
+            crictl: UNUSED
+          - runtime: Kubernetes-1.19
+            version: 1.19.4-00
+            crictl: UNUSED
+          - runtime: Kubernetes-1.20
+            version: 1.20.1-00
+            crictl: UNUSED
+        test:
+          - case: end-to-end
+            file: test/run-end-to-end.py
 
     steps:
-    - name: Checkout the head commit of the branch
-      uses: actions/checkout@v2
-      with:
-        persist-credentials: false
+      - name: Checkout the head commit of the branch
+        uses: actions/checkout@v2
+        with:
+          persist-credentials: false
+
+      - name: Setup Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - name: Install Python kubernetes dependency
+        run: |
+          python -m pip install --upgrade pip
+          pip install kubernetes
 
-    - name: Setup Python
-      uses: actions/setup-python@v2
-      with:
-       python-version: 3.8 
-    - name: Install Python kubernetes dependency
-      run: |
-        python -m pip install --upgrade pip
-        pip install kubernetes
+      - name: Download Agent container artifact
+        if: startsWith(github.event_name, 'pull_request')
+        uses: actions/download-artifact@v2
+        with:
+          name: agent.tar
+      - name: Download Controller container artifact
+        if: startsWith(github.event_name, 'pull_request')
+        uses: actions/download-artifact@v2
+        with:
+          name: controller.tar
 
-    - name: Download Agent container artifact
-      if: startsWith(github.event_name, 'pull_request')
-      uses: actions/download-artifact@v2
-      with:
-        name: agent.tar
-    - name: Download Controller container artifact
-      if: startsWith(github.event_name, 'pull_request')
-      uses: actions/download-artifact@v2
-      with:
-        name: controller.tar
+      - if: startsWith(matrix.kube.runtime, 'K3s')
+        name: Install K3s
+        env:
+          INSTALL_K3S_VERSION: ${{ matrix.kube.version }}
+        run: |
+          sudo curl -sfL https://get.k3s.io | sh -
+          sudo addgroup k3s-admin
+          sudo adduser $USER k3s-admin
+          sudo usermod -a -G k3s-admin $USER
+          sudo chgrp k3s-admin /etc/rancher/k3s/k3s.yaml
+          sudo chmod g+r /etc/rancher/k3s/k3s.yaml
+          sudo chmod 666 /etc/rancher/k3s/*
+          if [ "${{ matrix.kube.runtime }}" == "K3s-1.16" ]; then
```

Comment:
this block of code worked for me if it ran after `until kubectl get node ${HOSTNAME,,}` ... maybe the rancher device-plugins folder takes a little time to be created?

Reply:
Ah, useful. Thanks. Will move it.

I dislike adding the exception for the version but, short of making `K3s-1.16` an entirely distinct step, this does the trick.

However, the workflow isn't (heavily) commented and I wonder what future readers will make of the change? 

Should we create a document that attempts to document what it all (!) does?

I feel the need to note somewhere your discovery of the K3s 1.16 device-plugins issue and record that as part of this change history (other than in the commits). Thoughts? [I have written a document for the `./test/run-webhook.py` script for the same reason]

Reply:
> However, the workflow isn't (heavily) commented and I wonder what future readers will make of the change?
> Should we create a document that attempts to document what it all (!) does?

I 100% think we should have a page for each distro with detailed steps ... and any quirks we find.

While I was shoveling snow (the first time), I imagined refactoring all of the shell commands in the test workflow into actual script files one to set up k3s, one to set up microk8s, etc) ... and those scripts could serve as documented how-to.  That would feed the effort to move the python tests to bash as well.  It sounded like a great plan while I was shoveling.

Reply:
I like this idea alot!

When I was walking Freddie through the icy-slush this morning on the trail (:smile: ), I was thinking, if each step were more reproducible (script|container), it would be easier to transform the workflow into a script that builds e.g. VMs on Azure|GCP that could the be more easily introspected (GitHub workflows are reproducible across runs but very difficult to reproduce outside of GitHub; my K3s and MicroK8s VMs on GCE behave different than the tests even though I'm attempting to use the same scripts).

Reply:
> this block of code worked for me if it ran after `until kubectl get node ${HOSTNAME,,}` ... maybe the rancher device-plugins folder takes a little time to be created?

Fixed: https://github.com/deislabs/akri/pull/244/commits/7ab2d18d632d39280c48ef2d681bbfaebb951e9e